### PR TITLE
Follow markdown links into a new sidebar slot

### PR DESCRIPTION
**Pull request by larspohlmann**
## Summary

Click a `[link](other.md)` in the rendered preview and the linked file opens in a new sidebar slot. Works for relative paths, absolute file URLs, and `#fragment`-suffixed links (fragment is stripped).

When the linked file lives outside the per-file sandbox grant of the currently-open document, the app prompts once for folder-scoped access via NSOpenPanel; the bookmark is persisted so subsequent clicks under the same folder open instantly.

### How it works

- **Click capture** — `markdownobserver-link-intercept.js` (loaded via `BundledAssetLoader`) attaches a capture-phase click listener that detects `.md` / `.markdown` / `.mdown` anchors, prevents default navigation, and posts the resolved href to a `WKScriptMessageHandler`. Required because WKWebView silently drops `file://` link navigation for documents loaded with `loadHTMLString(html, baseURL: bundleURL)`.
- **URL resolution** — `MarkdownLinkResolver` strips the bundle prefix WKWebView added when resolving relative hrefs, drops fragments, and re-resolves against the currently-open document's directory. Reuses `FileRouting.isSupportedMarkdownFileURL` for extension matching.
- **Slot dispatch** — `.openLinkedFile(URL)` ⇒ `.requestFileOpen(FileOpenRequest(origin: .linkFollow, slotStrategy: .alwaysAppend))`.
- **Sandbox grant** — `WindowDocumentOpenCoordinator.openFileRequest` checks `LinkAccessGrantsStore` for an existing covering grant; if none, presents `LinkFollowAccessRequester`'s NSOpenPanel pre-pointed at the linked file's parent folder. `SecurityScopeResolver.linkAccessScopedAccessibleFileURL` activates the folder-scoped bookmark and returns a folder-relative URL the file loader can read.

### What's new

- `MarkdownLinkResolver` (model + tests)
- `markdownobserver-link-intercept.js` resource + `BundledAssetLoader.markdownLinkInterceptJavaScript`
- `OpenOrigin.linkFollow`, `DocumentSurfaceAction.openLinkedFile(URL)`
- `LinkAccessGrant` model + `LinkAccessGrantsStore` (mirrors `TrustedImageFoldersStore`, with deepest-match-wins resolution)
- `LinkFollowAccessRequester` service
- `SecurityScopeResolver.linkAccessScopedAccessibleFileURL` consults grant store after watch sessions
- 17 link-intercept regression tests + 9 grant-store tests

## Test plan

- [ ] Open a `.md` file via terminal `open` (per-file sandbox only). Click a relative `[link](sibling.md)` — expect the "Allow Link Following" panel pre-pointed at the parent folder. Grant access; the linked file opens in a new sidebar slot.
- [ ] Click another link to a different sibling under the same folder — expect no prompt; opens immediately.
- [ ] Click an external `https://` link — opens in default browser, no slot added.
- [ ] Click a `#fragment` anchor — scrolls in-page, no slot added.
- [ ] Click a `.png` link — does nothing (or hands off as before), no slot added.
- [ ] Quit + relaunch the app, click a link in the same folder — bookmark persisted, opens without prompt.
- [ ] Run unit tests: `xcodebuild test -project minimark.xcodeproj -scheme minimark -destination 'platform=macOS' -only-testing:minimarkTests` — all green.

### CODE REVIEW

File: minimark/Views/Window/Coordination/WindowDocumentOpenCoordinator.swift [hunk 1]
```diff
@@ -79,10 +83,27 @@ final class WindowDocumentOpenCoordinator {
     // MARK: - Open entry points
 
     func openFileRequest(_ request: FileOpenRequest) {
+        if request.origin == .linkFollow {
+            ensureLinkFollowAccess(forFiles: request.fileURLs)
+        }
         fileOpenCoordinator.open(request)
```
Comment (larspohlmann) [inline minimark/Views/Window/Coordination/WindowDocumentOpenCoordinator.swift:89]: **Bug: file open proceeds regardless of access grant outcome.**

`ensureLinkFollowAccess` calls `requestAccess`, which returns `false` when the user cancels the panel or picks a folder that doesn't contain the target file. But this return value is discarded — the file open proceeds unconditionally on the next line.

The sandbox will block the read, resulting in a new sidebar slot in an error/broken state with no clear feedback to the user about why.

Suggested fix — make `ensureLinkFollowAccess` return `Bool` and guard the open:

```swift
func openFileRequest(_ request: FileOpenRequest) {
    if request.origin == .linkFollow {
        guard ensureLinkFollowAccess(forFiles: request.fileURLs) else { return }
    }
    fileOpenCoordinator.open(request)
    callbacks.refreshWindowPresentation()
}
```

And change `ensureLinkFollowAccess` to return the result of `requestAccess`.

File: minimark/Models/LinkAccessGrant.swift [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    nonisolated var id: String {
+        folderPath
+    }
+
+    nonisolated var folderURL: URL {
+        URL(fileURLWithPath: folderPath)
+    }
+
+    init(folderURL: URL) {
```
Comment (larspohlmann) [inline minimark/Models/LinkAccessGrant.swift]: **`try?` silently swallows bookmark creation failures, creating a prompt loop.**

If `bookmarkData(options: .withSecurityScope)` fails (entitlement issue, URL doesn't support scopes, etc.), `bookmarkData` is set to `nil`. The grant is still persisted via `addLinkAccessGrant` → `insertingUnique`.

On the next link click:
1. `resolvedLinkAccessFolderURL` finds the grant but skips it (`bookmarkData` is `nil`)
2. `ensureLinkFollowAccess` sees no covering grant and re-prompts
3. New grant also gets `nil` bookmark data → same result

This is an infinite prompt loop.

Two fixes to consider:
1. Don't persist grants with `nil` bookmark data — `insertingUnique` could filter out entries with no bookmark, or `addLinkAccessGrant` could validate before persisting.
2. Have `requestAccess` check whether the grant was actually usable (e.g., verify `bookmarkData != nil` after creation) and return `false` if not. The caller should then abort the file open (see comment on `WindowDocumentOpenCoordinator`).

File: minimark/Views/Window/Coordination/WindowDocumentOpenCoordinator.swift [hunk 3]
```diff
@@ -79,10 +83,27 @@ final class WindowDocumentOpenCoordinator {
...
         callbacks.refreshWindowPresentation()
     }
 
+    /// Sandboxed apps only have access to files the user explicitly grants.
+    /// When following a markdown link to a sibling/descendant file, that file
+    /// usually isn't accessible until the user grants the parent folder. We
+    /// detect that here and present an NSOpenPanel to collect the grant —
+    /// which is then persisted so future link clicks under the same folder
+    /// don't re-prompt.
+    private func ensureLinkFollowAccess(forFiles fileURLs: [URL]) {
```
Comment (larspohlmann) [inline minimark/Views/Window/Coordination/WindowDocumentOpenCoordinator.swift]: This only checks `fileURLs.first`. Link follow always sends exactly one URL (set in `DocumentSurfaceViewModel:215`), so this is correct in practice. Consider adding a `precondition(fileURLs.count == 1)` or a comment documenting the single-URL assumption — the array type suggests multiple files are supported, which could mislead future callers.

File: minimark/Services/SecurityScopeResolver.swift [hunk 4]
```diff
@@ -141,6 +147,44 @@ final class SecurityScopeResolver {
...
+    /// Returns nil if no grant covers the URL or the bookmark cannot be
+    /// resolved.
+    func linkAccessScopedAccessibleFileURL(for fileURL: URL) -> URL? {
+        let normalizedFileURL = FileRouting.normalizedFileURL(fileURL)
+        guard let grantFolderURL = settingsStore.resolvedLinkAccessFolderURL(containing: normalizedFileURL) else {
+            return nil
+        }
+
+        if needsFolderTokenRefresh(for: grantFolderURL) {
+            context.folderToken?.endAccess()
```
Comment (larspohlmann) [inline minimark/Services/SecurityScopeResolver.swift:170]: This reuses `context.folderToken` which is also used by `ensureFolderWatchAccessIfNeeded` for folder watch sessions. If a document is being watched (folder A) and the user follows a link to a file in folder B, this code replaces the folder A token with folder B's token.

If the folder watch then delivers an event for folder A, `ensureFolderWatchAccessIfNeeded` (line 108) has an early return when `context.folderToken?.didStartAccess == true` — it would see the folder B token is active and skip re-activation, leaving the folder A file with a stale scope.

Whether this is an actual problem depends on whether `SecurityScopeResolver` is per-document (each sidebar slot has its own resolver) or shared. If per-document, the original document's resolver is unaffected. But the shared `context.folderToken` design makes this fragile — a comment or assertion documenting the per-instance assumption would be defensive.

File: minimark/Views/MarkdownWebView.swift [hunk 5]
```diff
@@ -591,6 +611,11 @@ struct MarkdownWebView: NSViewRepresentable {
             return .cancel
         }
 
+        private func currentDocumentDirectoryPath() -> String? {
```
Comment (larspohlmann) [inline minimark/Views/MarkdownWebView.swift:614]: This derives the document directory from `lastDocumentIdentity`, which is set to `fileURL?.standardizedFileURL.path` in `DocumentSurfaceViewModel:189`. This works correctly for normal file paths.

One edge case: if `documentIdentity` is ever `nil` (no file open), this returns `nil`, and `MarkdownLinkResolver.resolveMarkdownLink` returns `nil` for bundle-prefixed URLs when `documentDirectoryPath` is nil. Relative links would silently do nothing — this is correct behavior (can't resolve a relative link without a document context), but worth being aware of.

File: minimark/App/Resources/markdownobserver-link-intercept.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        if (e.defaultPrevented) return;
+        if (e.button !== 0) return;
+        if (e.metaKey || e.ctrlKey || e.shiftKey || e.altKey) return;
+        var a = e.target.closest && e.target.closest('a');
+        if (!a) return;
+        var rawHref = a.getAttribute('href');
+        if (!isMarkdownHref(rawHref)) return;
+        var resolvedHref = a.href;
+        if (!resolvedHref) return;
+        e.preventDefault();
```
Comment (larspohlmann) [inline minimark/App/Resources/markdownobserver-link-intercept.js:20]: The `isMarkdownHref` check strips both `#fragment` and `?query` before checking the extension, which is good. However, `resolvedHref = a.href` still includes the query string — it's the full resolved URL including `?query`. This means a link like `[x](file.md?param=value)` would be intercepted and the URL with query string would be sent to Swift.

`MarkdownLinkResolver` uses `URLComponents` to strip the fragment but doesn't strip the query string. The resulting `candidatePath` would include `?param=value` as part of the path, which would fail the `isSupportedMarkdownFileURL` check (the extension would be `md?param=value`... actually no, `.path` in URLComponents strips query). Let me reconsider — `bareURL.path` from `URLComponents.url` does strip query. So this is fine. Disregard.

**Review by larspohlmann — COMMENTED** [review]
## Review: Follow markdown links into a new sidebar slot

Overall the architecture is clean — the JS interception approach is well-justified (WKWebView silently dropping `file://` navigation for `loadHTMLString`-loaded documents is a real gotcha), the `MarkdownLinkResolver` is nicely testable, and the `LinkAccessGrantsStore` follows the established `TrustedImageFoldersStore` pattern faithfully with a meaningful improvement (deepest-match-wins resolution).

There are two bugs that should be fixed before merging:

### 1. File open proceeds after user denies access (Medium)

`WindowDocumentOpenCoordinator.openFileRequest` calls `ensureLinkFollowAccess` but ignores its result. If the user cancels the NSOpenPanel or picks a folder that doesn't contain the target file, `requestAccess` returns `false`, but `fileOpenCoordinator.open(request)` proceeds anyway. The sandbox will block the file read, resulting in a new sidebar slot that shows a broken/error state with no explanation to the user.

The fix is to make `ensureLinkFollowAccess` return a `Bool` and guard the file open on it.

### 2. Silent bookmark failure creates a prompt loop (Medium)

`LinkAccessGrant.init(folderURL:)` uses `try?` for bookmark creation, so failures produce a grant with `nil` `bookmarkData`. The grant is persisted and takes one of the 50 slots, but `resolvedLinkAccessFolderURL` skips entries with `nil` bookmark data. On the next link click, `ensureLinkFollowAccess` finds no covering grant and re-prompts. The new grant also gets `nil` bookmark data, perpetuating the cycle.

At minimum, `addLinkAccessGrant` should not persist grants with `nil` bookmark data (or should log the failure so it's diagnosable). More robustly, `requestAccess` should return `false` when the bookmark couldn't be created, and the caller should act on that.

### Test gaps

- **`SecurityScopeResolver.linkAccessScopedAccessibleFileURL`** — the most complex new code path (bookmark activation, folder token refresh, relative path construction) has no test coverage. Token swap edge cases in particular deserve a test.
- **`LinkFollowAccessRequester`** — the `folderContains` guard and the user-cancellation/non-containing-folder paths are untested. The `pickFolder` closure is injectable, which is good, but no test exercises it.

### Minor observations

- `ensureLinkFollowAccess` only checks `fileURLs.first` — harmless since link follow always sends one URL, but worth a `precondition` or at least a comment documenting the assumption.
- `resolvedLinkAccessFolderURL` is called twice per link-follow file open (once in `ensureLinkFollowAccess` for the existence check, once in `SecurityScopeResolver` to activate scope). The bookmark resolution is non-trivial. Not a bug, but a candidate for optimization if link-following becomes a hot path.

File: minimark/Views/Window/Coordination/WindowDocumentOpenCoordinator.swift (hunk 1 above)
Comment (larspohlmann) [reply minimark/Views/Window/Coordination/WindowDocumentOpenCoordinator.swift:89]: Fixed. `ensureLinkFollowAccess` now returns `Bool`; `openFileRequest` guards on it and aborts the open when access wasn't obtained. `requestAccess` already returned `Bool` — wired it through and additionally re-checks `resolvedLinkAccessFolderURL` so a panel-cancel, mismatched-folder pick, or failed bookmark all surface as `false`.

File: minimark/Models/LinkAccessGrant.swift (hunk 2 above)
Comment (larspohlmann) [reply minimark/Models/LinkAccessGrant.swift]: Fixed. Moved bookmark creation out of `LinkAccessGrant` into `LinkAccessGrantsStore.addLinkAccessGrant`, which now uses the injected `BookmarkRefreshing.create` and bails to a no-op when creation throws. `LinkAccessGrantHistory.insertingUnique` now takes the bookmark data explicitly and skips the insert when it's `nil`, so a phantom grant can't get persisted. `LinkFollowAccessRequester.requestAccess` then re-checks `resolvedLinkAccessFolderURL` after the add and returns `false` on failure, which the coordinator (see thread on `WindowDocumentOpenCoordinator.swift:89`) uses to abort the open. New tests cover both branches: `insertingUniqueSkipsEntriesWithoutBookmark` and `addLinkAccessGrantSkipsWhenBookmarkCreationFails`.

File: minimark/Views/Window/Coordination/WindowDocumentOpenCoordinator.swift (hunk 3 above)
Comment (larspohlmann) [reply minimark/Views/Window/Coordination/WindowDocumentOpenCoordinator.swift]: Added `precondition(fileURLs.count == 1)` and a doc-comment on `ensureLinkFollowAccess` calling out the single-URL contract (with a pointer to `DocumentSurfaceViewModel.openLinkedFile(_:)`).

File: minimark/Services/SecurityScopeResolver.swift (hunk 4 above)
Comment (larspohlmann) [reply minimark/Services/SecurityScopeResolver.swift:170]: Verified: `SecurityScopeResolver` is per-`DocumentStore` (built in `SidebarDocumentController`'s store factory), so each slot has its own `context.folderToken` — a link-follow's grant token can't stomp another slot's watched-folder token. Within a single resolver, `effectiveAccessibleFileURL` calls `folderScopedAccessibleFileURL` (watch-session path) before `linkAccessScopedAccessibleFileURL`, so a watched file never reaches the link-access branch. Added a doc-comment on `linkAccessScopedAccessibleFileURL` recording both invariants.

File: minimark/Views/MarkdownWebView.swift (hunk 5 above)
Comment (larspohlmann) [reply minimark/Views/MarkdownWebView.swift:614]: Acknowledged — leaving as-is. The nil-`documentIdentity` case correctly produces no-op for relative links (no document context to resolve against), and absolute `file://` links still resolve.

File: minimark/App/Resources/markdownobserver-link-intercept.js (hunk 6 above)
Comment (larspohlmann) [reply minimark/App/Resources/markdownobserver-link-intercept.js:20]: Confirmed — no action needed.
